File: 204.count-primes.js
/*
 * @lc app=leetcode id=204 lang=javascript
 *
 * [204] Count Primes
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var countPrimes = function(n) {
  // NEW solution for +180IQ:
  // Also known as Sieve of Eratosthenes
  let sieve = new Array(n).fill(true)
  let primes = []
  for (let i = 2; i < n ** .5; i++) {
    if (sieve[i]) {
      for (let j = i ** 2; j < n; j += i) {
        sieve[j] = false
      }
    }
  }
  // NOTE: 0 and 1 are not prime
  return sieve.reduce((acc, v, k) =>
    v && k > 1 ? acc += 1 : acc, 0
  )


  // OLD solution for chumps:
  // let isPrime = (num) => {
  //   for (let j of cache) {
  //     if (num % j === 0) return false
  //     if (j >= num ** .5) break
  //   }
  //   return true
  // }

  // if (n <= 2) return 0
  // if (n <= 3) return 1
  // if (n <= 5) return 2
  // if (n <= 6) return 3

  // let cache = new Set([2, 3])

  // for (let i = 6; i <= n; i += 6) {
  //   if (isPrime(i-1)) cache.add(i-1)
  //   if (i >= (n-1)) break
  //   if (isPrime(i+1)) cache.add(i+1)
  // }

  // return cache.size
};
// @lc code=end

File: cpp/204.count-primes.cpp
/*
 * @lc app=leetcode id=204 lang=cpp
 *
 * [204] Count Primes
 */

// @lc code=start
class Solution
{
public:
    // unordered_set<int> primesCache;
    // bool isPrime(int num)
    // {
    //     // check primes cache first
    //     for (int prime : primesCache)
    //     {
    //         if (num % prime == 0)
    //         {
    //             return false;
    //         }
    //     }
    //     // exhaustively check from 2-sqrt(num)
    //     for (int i = 2; i < pow(num, .5) + 1; i++)
    //     {
    //         if (num % i == 0)
    //         {
    //             return false;
    //         }
    //     }
    //     primesCache.insert(num);
    //     return true;
    // }
    int countPrimes(int n)
    {
        // brute force
        // iterate from [2, n), counting all primes
        // create helper function for isPrime
        // cache primes for speed up
        // O(n^2) time O(k) space
        // int primesCounter = 1;
        // primesCache.insert(2);
        // for (int i = 2; i < n; i++)
        // {
        //     if (isPrime(i))
        //         primesCounter++;
        // }
        // return primesCounter;

        // sieve of eratosthenes approach
        // O(n^2) time O(n) space
        vector<bool> sieve(n);
        fill(sieve.begin(), sieve.end(), false);

        for (int i = 2; i < n; i++)
        {
            // unchecked/prime
            if (!sieve[i])
            {
                // mark every next multiple of i
                int j = i + i;
                while (j < n)
                {
                    sieve[j] = true;
                    j += i;
                }
            }
        }
        // count false/prime values
        int primesCount = 0;
        for (int i = 2; i < n; i++)
        {
            if (!sieve[i])
                primesCount++;
        }
        return primesCount;
    }
};
// @lc code=end
